contracts/cw20-merkle-airdrop/src/contract.rs

```diff
@@ -0,0 +1,2158 @@
+            .as_slice()
+            .try_into()
+            .map_err(|_| ContractError::WrongLength {})
+    })?;
+
+    let mut root_buf: [u8; 32] = [0; 32];
+    hex::decode_to_slice(merkle_root, &mut root_buf)?;
+    if root_buf != hash {
+        return Err(ContractError::VerificationFailed {});
+    }
+
+    let credits_address = match config.credits_address {
+        Some(addr) => addr,
+        None => return Err(ContractError::CreditsAddress {}),
+    };
+
+    // Update claim index to the current stage
+    CLAIM.save(deps.storage, (proof_addr, stage), &true)?;
+
+    // Update total claimed to reflect
+    let mut claimed_amount = STAGE_AMOUNT_CLAIMED.load(deps.storage, stage)?;
+    claimed_amount += amount;
+    STAGE_AMOUNT_CLAIMED.save(deps.storage, stage, &claimed_amount)?;
+
+    // we stop all airdrops at the date of expiration, and we use the very same date
+    // as a start timestamp for vesting. if expiration is not set, vesting will not work.
+    let vesting_start_time =
+        match expiration {
+            Expiration::AtHeight(_) => {
+                return Err(ContractError::Vesting {
+                    description: "Vesting must be scheduled from timestamp, not block height"
+                        .to_string(),
+                })
+            }
+            Expiration::Never {} => return Err(ContractError::Vesting {
+                description:
+                    "Vesting must be scheduled from some timestamp, but there wasn't any provided"
+                        .to_string(),
+            }),
+            Expiration::AtTime(timestamp) => timestamp,
+        };
+
+    let transfer_message = Cw20Contract(credits_address.clone())
+        .call(Cw20ExecuteMsg::Transfer {
+            recipient: info.sender.to_string(),
+            amount,
+        })
+        .map_err(ContractError::Std)?;
+    let vesting_message = WasmMsg::Execute {
+        contract_addr: credits_address.to_string(),
+        msg: to_binary(&AddVesting {
+            address: info.sender.to_string(),
+            amount,
+            start_time: vesting_start_time.nanos(),
+            duration: VESTING_DURATION,
+        })?,
+        funds: vec![],
+    };
+    let res = Response::new()
+        .add_message(transfer_message)
+        .add_message(vesting_message)
+        .add_attributes(vec![
+            attr("action", "claim"),
+            attr("stage", stage.to_string()),
+            attr("address", info.sender.to_string()),
+            attr("amount", amount),
+        ]);
+    Ok(res)
+}
+
+pub fn execute_withdraw_all(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
+    // authorize owner
+    let cfg = CONFIG.load(deps.storage)?;
+    let owner = cfg.owner.ok_or(ContractError::Unauthorized {})?;
+    if info.sender != owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    if !STAGE_PAUSED.load(deps.storage, 1)?
```

Comment:
What is `1`?

Reply:
> also an `is_paused = STAGE_PAUSED.load(deps.storage, stage)` check is required here

I am not even sure if we are going to use pause functionality at all. I personally planned to remove it.

> also why we have stage as a param for `execute_claim`

because that's exactly how it has been in original contract, I didn't change this part.

Reply:
imo can be done in a separate PR. marking it with this msg to create a list of such complaints a bit later

Reply:
Closing since I have removed stages anyway.